### DISCUSSION TITLE
Speeding up startup time (with packer)

### DISCUSSION THREAD

**Discussion by oschrenk** [scalameta/nvim-metals#502] [Q&A] [answered]
I'm trying to optimize my nvim configuration, focussing currently on my startup time. Visualizing  `nvim --startuptime startup.log`  with [1] shows
```
  1        startup: 102.3
  2 event                  time percent plot
  3 packer_compiled.lua   30.36   29.66 ██████████████████████████
  4 init.lua              18.01   17.60 ███████████████▍
  5 config.metals_config  16.75   16.37 ██████████████▍
  6 metals                13.73   13.41 ███████████▊
  7 impatient             10.53   10.29 █████████
  8 treesitter.lua         8.34    8.14 ███████▏
  9 loading rtp plugins    7.80    7.62 ██████▋
 10 loading packages       7.80    7.62 ██████▋
 11 metals.config          4.64    4.53 ████
 12 metals.commands        4.50    4.40 ███▉

```

That means that 16.75 + 13.73 + 4.64 + 4.50 = 39.62ms, roughly 38% of the startup time of a new buffer is spent on loading metals.  
I use mainly the minimal configuration from [2]

```
use {
  'scalameta/nvim-metals',
   requires = {
    'hrsh7th/cmp-nvim-lsp'
  },
  config = [[require('config.metals_config')]],
}

```

```
local metals_config = require("metals").bare_config()

metals_config.settings = {
  showImplicitArguments = true,
  excludedPackages = { "akka.actor.typed.javadsl", "com.github.swagger.akka.javadsl" },
}
metals_config.init_options.statusBarProvider = "on"

metals_config.capabilities = require("cmp_nvim_lsp").default_capabilities()

vim.cmd([[autocmd FileType scala,sbt lua require("metals").initialize_or_attach({})]])

local function map(mode, lhs, rhs, opts)
  local options = { noremap = true }
  if opts then
    options = vim.tbl_extend("force", options, opts)
  end
  vim.api.nvim_set_keymap(mode, lhs, rhs, options)
end

map("n", "gD", "<cmd>lua vim.lsp.buf.definition()<CR>")
map("n", "K", "<cmd>lua vim.lsp.buf.hover()<CR>")
map("n", "gi", "<cmd>lua vim.lsp.buf.implementation()<CR>")
map("n", "gr", "<cmd>lua vim.lsp.buf.references()<CR>")
map("n", "gds", "<cmd>lua vim.lsp.buf.document_symbol()<CR>")
map("n", "gws", "<cmd>lua vim.lsp.buf.workspace_symbol()<CR>")
map("n", "<leader>sh", [[<cmd>lua vim.lsp.buf.signature_help()<CR>]])
map("n", "<leader>rn", "<cmd>lua vim.lsp.buf.rename()<CR>")
map("n", "<leader>f", "<cmd>lua vim.lsp.buf.format{async = true }<CR>")
```

Am I interpreting the data wrong? Any ideas how to improve my setup?

[1] https://github.com/dstein64/vim-startuptime
[2] https://github.com/scalameta/nvim-metals/discussions/39

**Comment by ckipp01** [comment]
Hey @oschrenk thanks for opening this! Honestly, this is a great question, and one I haven't dove into a ton. I typically don't care unless something becomes noticeably slow for me, and I guess I haven't hit that yet 😆. So I'm sure this can be improved, but I haven't spent a lot of time on it. The example setup for sure isn't optimized, but more just an example. So let's see if we can clean all this up.

My first reaction is that the way you have `config` setup is that it's calling this on every open, and in reality we don't want to. We only want to when we're actually opening a Scala file. One way we can do that is to ensure that all the Metals config stuff is done more or less lazily by moving everything right into the autocmd. Again, haven't tried or played around with this yet, but I'll try to later.

Typically we have something like this

```lua
  api.nvim_create_autocmd("FileType", {
    pattern = { "scala", "sbt", "java" },
    callback = function()
      require("metals").initialize_or_attach(metals_config)
    end,
    group = nvim_metals_group,
  })
```

When in reality we could refactor this a bit to something like this:

```lua
  api.nvim_create_autocmd("FileType", {
    pattern = { "scala", "sbt", "java" },
    callback = function()
      local config = metals_setup()
      require("metals").initialize_or_attach(config)
    end,
    group = nvim_metals_group,
  })
```

So the `metals_setup()` wouldn't actually be called until we actually need it. Then there'd be no `require("metals")` anything util that point. I'd recommend playing around with this a bit, and please do report back!

**Reply by oschrenk** [reply]
Your idea of lazy loading led me to the feature of packer to only load thing based on filetype, in this case ` ft = {'scala', 'sbt' },`:

```
  use {
    'scalameta/nvim-metals',
    requires = {
      'hrsh7th/cmp-nvim-lsp'
    },
    ft = {'scala', 'sbt' },
    config = [[require('config.metals_config')]],
  }

```

**Comment by oschrenk** [comment, marked as answer]
Your idea of lazy loading led me to the feature of packer to only load thing based on filetype, in this case ` ft = {'scala', 'sbt' },`:

```
  use {
    'scalameta/nvim-metals',
    requires = {
      'hrsh7th/cmp-nvim-lsp'
    },
    ft = {'scala', 'sbt' },
    config = [[require('config.metals_config')]],
  }

```

**Reply by oschrenk** [reply]
Now it does not load metals during general startup. For me that's good enough.
